src/Orleans/Async/Cancellation/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.Async
+{
+    /// <summary>
+    /// Contains list of cancellation token source corresponding to the tokens
+    /// passed to the related grain activation.
+    /// </summary>
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private readonly Lazy<TraceLogger> _logger = new Lazy<TraceLogger>(() =>
+            TraceLogger.GetLogger("CancellationSourcesExtension", TraceLogger.LoggerType.Application));
+        private static readonly Interner<Guid, GrainCancellationTokenSource> _cancellationTokenSources;
```

Comment:
Application code will hold GTCS (the Token Source) and will be able to Cancel via it.
Since we fully implement GCTS (it is Orleans's GTCS, not the standard one), we can do what ever we want upon Cancel.

I think you are still thinking in terms on the old implementation that used standard Token and standard Token Source. But once we moved away to Orleans ones, we have the full freedom to implement it anyway we want.